Source/Particles/ParticleBoundaryBuffer.cpp

```diff
@@ -0,0 +1,223 @@
+            pp_species.query("scrape_zlo", m_do_boundary_buffer[4][ispecies]);
+            pp_species.query("scrape_zhi", m_do_boundary_buffer[5][ispecies]);
+#endif
+#ifdef AMREX_USE_EB
+            pp_species.query("scrape_eb", m_do_boundary_buffer[AMREX_SPACEDIM*2][ispecies]);
+#endif
+        }
+    }
+}
+
+void ParticleBoundaryBuffer::printNumParticles () const {
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+    {
+        for (int iside = 0; iside < 2; ++iside)
+        {
+            auto& buffer = m_particle_containers[2*idim+iside];
+            for (int i = 0; i < numSpecies(); ++i)
+            {
+                int np = buffer[i].isDefined() ? buffer[i].TotalNumberOfParticles(false) : 0;
+                amrex::Print() << "Species " << getSpeciesNames()[i] << " has "
+                               << np << " particles in the boundary buffer "
+                               << " for side " << iside << " of dim " << idim << "\n";
+            }
+        }
+    }
+#ifdef AMREX_USE_EB
+    auto& buffer = m_particle_containers[2*AMREX_SPACEDIM];
+    for (int i = 0; i < numSpecies(); ++i)
+    {
+        int np = buffer[i].isDefined() ? buffer[i].TotalNumberOfParticles(false) : 0;
+        amrex::Print() << "Species " << getSpeciesNames()[i] << " has "
+                       << np << " particles in the EB boundary buffer \n";
+    }
+#endif
+}
+
+void ParticleBoundaryBuffer::clearParticles () {
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        auto& buffer = m_particle_containers[i];
+        if (buffer[i].isDefined())
+        {
+            buffer[i].clearParticles();
+        }
+    }
+}
+
+void ParticleBoundaryBuffer::gatherParticles (MultiParticleContainer& mypc,
+                                              const amrex::Vector<const amrex::MultiFab*>& distance_to_eb) {
+    using SrcData = WarpXParticleContainer::ParticleTileType::ConstParticleTileDataType;
+    using DstData = ParticleBuffer::BufferType<amrex::PinnedArenaAllocator>::ParticleTileType::ParticleTileDataType;
+    using PIter = amrex::ParConstIter<0,0,PIdx::nattribs>;
+    const auto& warpx_instance = WarpX::GetInstance();
+    const amrex::Geometry& geom = warpx_instance.Geom(0);
+    auto plo = geom.ProbLoArray();
+    auto phi = geom.ProbHiArray();
+    auto dxi = geom.InvCellSizeArray();
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+    {
+        if (geom.isPeriodic(idim)) continue;
+        for (int iside = 0; iside < 2; ++iside)
+        {
+            auto& buffer = m_particle_containers[2*idim+iside];
+            for (int i = 0; i < numSpecies(); ++i)
+            {
+                if (!m_do_boundary_buffer[2*idim+iside][i]) continue;
+                const auto& pc = mypc.GetParticleContainer(i);
+                if (!buffer[i].isDefined())
+                {
+                    buffer[i] = ParticleBuffer::getTmpPC<amrex::PinnedArenaAllocator>(&pc);
+                    buffer[i].AddRealComp(false);  // for timestamp
```

Comment:
Do you have a specific preference for using the simulation time rather than the timestep? We will use this to build a dataframe of the total charge absorbed at each timestep by grouping particles scraped at the same step (which we will then translate to a current by dividing by `dt`). I'm thinking we might run into some issues with floating point presicion going from C++ to Python causing that grouping to be off so having a scraped timestep is slightly better for us (we could of course convert the simulation time back to a timestep so not a big deal at all).

Reply:
No real reason, just the first thing that came to mind. I have changed it to the time step instead.

Reply:
Thanks!